File: WordTarget/WordTargetCore/LinkTypes.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace WordTargetCore
{
    public class LinkTypes
    {
        public class LetterCounter
        {
            private Dictionary<char, int> counter = new Dictionary<char, int>();

            public void AddLetter(char letter)
            {
                if (counter.ContainsKey(letter))
                {
                    counter[letter]++;
                } else
                {
                    counter[letter] = 1;
                }
            }

            public int CountFor(char letter)
            {
                if (counter.ContainsKey(letter))
                {
                    return counter[letter];
                } else
                {
                    return 0;
                }
            }

            public override bool Equals(object obj)
            {
                if (obj is LetterCounter other) {
                    if (counter.Keys.Count != other.counter.Keys.Count)
                    {
                        return false;
                    }
                    foreach (KeyValuePair<char, int> kvp in counter)
                    {
                        if (!other.counter.ContainsKey(kvp.Key) || other.counter[kvp.Key] != kvp.Value) return false;
                    }
                }
                else
                {
                    return false;
                }

                return true;
            }

            public override int GetHashCode()
            {
                return -837799269 + EqualityComparer<Dictionary<char, int>>.Default.GetHashCode(counter);
            }

            public override string ToString()
            {
                string output = "[";
                foreach (KeyValuePair<char, int> kvp in counter)
                {
                    output += string.Format("{0}:{1}", kvp.Key, kvp.Value);
                    output += ";";
                }
                if (output.Length > 1) {
                    output = output.Remove(output.Length - 1, 1);
                }
                output += "]";
                return output;
            }
        }

        public static LetterCounter CountLetters(string text)
        {
            LetterCounter counter = new LetterCounter();
            foreach (char c in text)
            {
                counter.AddLetter(c);
            }
            return counter;
        }

        public static bool IsAnagram(string textA, string textB)
        {
            return CountLetters(textA).Equals(CountLetters(textB));
        }

        public static bool IsOneLetterChange(string textA, string textB)
        {
            if (textA.Length != textB.Length)
            {
                return false;
            }

            bool firstLetterChange = false;
            for (int i = 0; i < textA.Length; i++)
            {
                if (textA[i] != textB[i])
                {
                    if (firstLetterChange)
                    {
                        return false;
                    } 
                    else
                    {
                        firstLetterChange = true;
                    }
                
                }
            }
            return firstLetterChange;
        }

        public static bool IsOneLetterAdd(string textA, string textB)
        {
            if (textA.Length - textB.Length != -1)
            {
                return false;
            }
            
            bool firstLetterAdded = false;
            for (int i = 0; i < textA.Length; i++)
            {
                int offset = firstLetterAdded ? 1 : 0;
                if (textA[i] != textB[i + offset])
                {
                    if (firstLetterAdded)
                    {
                        return false;
                    }
                    else
                    {
                        firstLetterAdded = true;
                        i--;
                    }
                }
            }
            return true;
        }

        public static bool IsOneLetterRemove(string textA, string textB)
        {
            return IsOneLetterAdd(textB, textA);
        }
    }

}

File: word-target/word-target-core/src/main/java/it/carcassi/wordtarget/core/LinkTypes.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package it.carcassi.wordtarget.core;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;
import static it.carcassi.wordtarget.core.LetterCounter.*;

/**
 *
 * @author Matteo
 */
public class LinkTypes {

    public static boolean isAnagram(String textA, String textB) {
        return isAnagram(Word.of(textA), Word.of(textB));
    }
    
    public static boolean isAnagram(Word wordA, Word wordB) {
        if (wordA.getText().length() != wordB.getText().length()) {
            return false;
        }

        if (wordA.equals(wordB)) {
            return false;
        }

        return countLetters(wordA).equals(countLetters(wordB));
    }

    public static boolean isOneLetterChange(String textA, String textB) {
        if (textA.length() != textB.length()) {
            return false;
        }

        boolean firstLetterChange = false;
        for (int i = 0; i < textA.length(); i++) {
            if (textA.charAt(i) != textB.charAt(i)) {
                if (firstLetterChange) {
                    return false;
                } else {
                    firstLetterChange = true;
                }

            }
        }
        return firstLetterChange;
    }

    public static boolean isOneLetterAdd(String textA, String textB) {
        if (textA.length() - textB.length() != -1) {
            return false;
        }

        boolean firstLetterAdded = false;
        for (int i = 0; i < textA.length(); i++) {
            int offset = firstLetterAdded ? 1 : 0;
            if (textA.charAt(i) != textB.charAt(i + offset)) {
                if (firstLetterAdded) {
                    return false;
                } else {
                    firstLetterAdded = true;
                    i--;
                }
            }
        }
        return true;
    }

    public static boolean isOneLetterRemove(String textA, String textB) {
        return isOneLetterAdd(textB, textA);
    }

    public static boolean isOneLetterAddOrRemove(String textA, String textB) {
        return isOneLetterAdd(textA, textB) || isOneLetterRemove(textA, textB);
    }
    
    public static LinkType calculateLinkType(Word word1, Word word2) {
        if (LinkTypes.isOneLetterAddOrRemove(word1.toString(), word2.toString())) {
            return LinkType.OneLetterAddOrRemove;
        }
        if (LinkTypes.isOneLetterChange(word1.toString(), word2.toString())) {
            return LinkType.OneLetterChange;
        }
        if (LinkTypes.isAnagram(word1.toString(), word2.toString())) {
            return LinkType.Anagram;
        }
        else {
            return null;
        }
    }

//    public static LinkType fromString(String token) {
//        switch (token) {
//            case "OneLetterChange":
//                return LinkType.OneLetterChange;
//            case "OneLetterAddOrRemove":
//                return LinkType.OneLetterAddOrRemove;
//            case "Anagram":
//                return LinkType.Anagram;
//            case "Synonym":
//                return LinkType.Synonym;
//            case "Antonym":
//                return LinkType.Antonym;
//            case "WordAssociation":
//                return LinkType.WordAssociation;
//            default:
//                throw new Exception("The string " + token + " does not correspond to a valid link type");
//        }
//    }
}
